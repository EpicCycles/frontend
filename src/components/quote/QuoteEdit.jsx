/* eslint-disable react/no-access-state-in-setstate,react/destructuring-assignment */
import React from 'react';
import * as PropTypes from 'prop-types';
import { updateObject } from '../../helpers/utils';
import { checkForChangesAllFields, updateModel } from '../app/model/helpers/model';
import { quoteFields, quoteFieldsNoBike } from './helpers/display';
import EditModelPage from '../app/model/EditModelPage';
import ModelEditIcons from '../app/model/ModelEditIcons';

class QuoteEdit extends React.Component {
  state = {
    quote: updateObject(this.props.quote),
    persistedQuote: this.props.quote,
  };

  static getDerivedStateFromProps(props, state) {
    // Any time the current user changes,
    // Reset any parts of state that are tied to that user.
    // In this simple example, that's just the email.
    if (checkForChangesAllFields(quoteFields, props.quote, state.persistedQuote)) {
      return {
        quote: updateObject(props.quote),
        persistedQuote: props.quote,
      };
    }
    return null;
  }

  handleInputChange = (fieldName, input) => {
    let { quote } = this.state;
    quote = updateModel(quote, quoteFields, fieldName, input);
    this.setState({ quote });
  };

  onClickReset = () => {
    const { persistedQuote } = this.state;
    const quote = updateObject(persistedQuote);
    this.setState({ quote });
  };

  render() {
    const { quote, persistedQuote } = this.state;
    const fields = quote.bike ? quoteFields : quoteFieldsNoBike;
    const {
      componentKey,
      customers,
      bikes,
      frames,
      brands,
      users,
      saveQuote,
      archiveQuote,
    } = this.props;
    return (
      <div className="fit-content">
        <EditModelPage
          model={quote}
          persistedModel={persistedQuote}
          modelFields={fields}
          onChange={this.handleInputChange}
          showReadOnlyFields
          customers={customers}
          bikes={bikes}
          brands={brands}
          frames={frames}
          users={users}
        />
        <div className="align_right">
          <ModelEditIcons
            componentKey={componentKey}
            model={quote}
            modelSave={saveQuote}
            modelDelete={archiveQuote}
            modelReset={this.onClickReset}
          />
        </div>
      </div>
    );
  }
}

QuoteEdit.defaultProps = {
  quote: {},
};
QuoteEdit.propTypes = {
  quote: PropTypes.object,
  createQuote: PropTypes.func,
  archiveQuote: PropTypes.func.isRequired,
  saveQuote: PropTypes.func.isRequired,
  componentKey: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  bikes: PropTypes.array.isRequired,
  brands: PropTypes.array.isRequired,
  customers: PropTypes.array.isRequired,
  frames: PropTypes.array.isRequired,
  users: PropTypes.array.isRequired,
};
export default QuoteEdit;
