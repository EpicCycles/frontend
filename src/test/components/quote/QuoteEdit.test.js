import React from 'react';
import toJson from 'enzyme-to-json';

import QuoteEdit from '../../../components/quote/QuoteEdit';

describe('QuoteEdit tests', () => {
  const quote = {
    id: 16,
    add_date: '2018-07-04T13:02:09.988286+01:00',
    upd_date: '2018-07-04T13:02:09.988343+01:00',
  };
  it('renders the form text correctly with quote', () => {
    const input = shallow(<QuoteEdit quote={quote} />);
    expect(toJson(input)).toMatchSnapshot();
  });
  it('renders the form text correctly with no quote', () => {
    const input = shallow(<QuoteEdit />);
    expect(toJson(input)).toMatchSnapshot();
  });
});
