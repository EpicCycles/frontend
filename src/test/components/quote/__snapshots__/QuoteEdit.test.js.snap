// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuoteEdit tests renders the form text correctly with no quote 1`] = `
<div
  className="fit-content"
>
  <EditModelPage
    model={Object {}}
    modelFields={
      Array [
        Object {
          "displaySize": 40,
          "fieldName": "quote_desc",
          "header": "Description",
          "maxLength": 60,
          "maxWidth": "250px",
          "required": true,
          "type": "text",
        },
        Object {
          "fieldName": "customer",
          "header": "Customer",
          "readOnly": true,
          "type": "customer",
        },
        Object {
          "fieldName": "quote_status",
          "header": "Status",
          "readOnly": true,
          "selectList": Array [
            Object {
              "name": "New",
              "value": "1",
            },
            Object {
              "name": "Issued",
              "value": "2",
            },
            Object {
              "name": "Archived",
              "value": "3",
            },
            Object {
              "name": "Order Created",
              "value": "4",
            },
          ],
          "type": "select_one",
        },
        Object {
          "fieldName": "club_member",
          "header": "Club Member?",
          "type": "checkbox",
        },
        Object {
          "fieldName": "version",
          "header": "Version",
          "readOnly": true,
          "type": "number",
        },
        Object {
          "fieldName": "created_by",
          "header": "Created By",
          "readOnly": true,
          "type": "user",
        },
        Object {
          "fieldName": "created_date",
          "header": "Date Created",
          "readOnly": true,
          "type": "date_time",
        },
        Object {
          "fieldName": "upd_date",
          "header": "Updated",
          "readOnly": true,
          "type": "date_time",
        },
        Object {
          "fieldName": "issued_date",
          "header": "Date Issued",
          "readOnly": true,
          "type": "date_time",
        },
        Object {
          "displaySize": 7,
          "fieldName": "quote_price",
          "header": "Quote £",
          "maxLength": 10,
          "synonyms": Array [],
          "type": "currency",
        },
        Object {
          "displaySize": 6,
          "fieldName": "calculated_price",
          "header": "Total £",
          "maxLength": 10,
          "readOnly": true,
          "type": "currency",
        },
      ]
    }
    onChange={[Function]}
    persistedModel={Object {}}
    showReadOnlyFields={true}
  />
  <div
    className="align_right"
  >
    <ModelEditIcons
      componentKey="new"
      model={Object {}}
      modelReset={[Function]}
    />
  </div>
</div>
`;

exports[`QuoteEdit tests renders the form text correctly with quote 1`] = `
<div
  className="fit-content"
>
  <EditModelPage
    model={
      Object {
        "add_date": "2018-07-04T13:02:09.988286+01:00",
        "id": 16,
        "upd_date": "2018-07-04T13:02:09.988343+01:00",
      }
    }
    modelFields={
      Array [
        Object {
          "displaySize": 40,
          "fieldName": "quote_desc",
          "header": "Description",
          "maxLength": 60,
          "maxWidth": "250px",
          "required": true,
          "type": "text",
        },
        Object {
          "fieldName": "customer",
          "header": "Customer",
          "readOnly": true,
          "type": "customer",
        },
        Object {
          "fieldName": "quote_status",
          "header": "Status",
          "readOnly": true,
          "selectList": Array [
            Object {
              "name": "New",
              "value": "1",
            },
            Object {
              "name": "Issued",
              "value": "2",
            },
            Object {
              "name": "Archived",
              "value": "3",
            },
            Object {
              "name": "Order Created",
              "value": "4",
            },
          ],
          "type": "select_one",
        },
        Object {
          "fieldName": "club_member",
          "header": "Club Member?",
          "type": "checkbox",
        },
        Object {
          "fieldName": "version",
          "header": "Version",
          "readOnly": true,
          "type": "number",
        },
        Object {
          "fieldName": "created_by",
          "header": "Created By",
          "readOnly": true,
          "type": "user",
        },
        Object {
          "fieldName": "created_date",
          "header": "Date Created",
          "readOnly": true,
          "type": "date_time",
        },
        Object {
          "fieldName": "upd_date",
          "header": "Updated",
          "readOnly": true,
          "type": "date_time",
        },
        Object {
          "fieldName": "issued_date",
          "header": "Date Issued",
          "readOnly": true,
          "type": "date_time",
        },
        Object {
          "displaySize": 7,
          "fieldName": "quote_price",
          "header": "Quote £",
          "maxLength": 10,
          "synonyms": Array [],
          "type": "currency",
        },
        Object {
          "displaySize": 6,
          "fieldName": "calculated_price",
          "header": "Total £",
          "maxLength": 10,
          "readOnly": true,
          "type": "currency",
        },
      ]
    }
    onChange={[Function]}
    persistedModel={
      Object {
        "add_date": "2018-07-04T13:02:09.988286+01:00",
        "id": 16,
        "upd_date": "2018-07-04T13:02:09.988343+01:00",
      }
    }
    showReadOnlyFields={true}
  />
  <div
    className="align_right"
  >
    <ModelEditIcons
      componentKey="new"
      model={
        Object {
          "add_date": "2018-07-04T13:02:09.988286+01:00",
          "id": 16,
          "upd_date": "2018-07-04T13:02:09.988343+01:00",
        }
      }
      modelReset={[Function]}
    />
  </div>
</div>
`;
